src/Package/Impl/ProjectSystem/Configuration/ProjectConfigurationSettingsProvider.cs

```diff
@@ -0,0 +1,124 @@
+                if (_settings.TryGetValue(projectFolder, out access)) {
+                    access.UseCount++;
+                } else {
+                    var settings = await OpenCollectionAsync(projectFolder, propertes);
+                    access = new SettingsAccess(this, projectFolder, settings);
+                    _settings[projectFolder] = access;
+                }
+
+            } finally {
+                _semaphore.Release();
+            }
+
+            return access;
+        }
+
+        private async Task<ConfigurationSettingCollection> OpenCollectionAsync(string projectFolder, IRProjectProperties properties) {
+            var settings = new ConfigurationSettingCollection();
+            var settingsFilePath = await GetSettingsFilePathAsync(projectFolder, properties);
+            if (!string.IsNullOrEmpty(settingsFilePath)) {
+                settings.Load(settingsFilePath.MakeAbsolutePathFromRRelative(projectFolder));
+            }
+            return settings;
+        }
+
+        private async Task<string> GetSettingsFilePathAsync(string projectFolder, IRProjectProperties properties) {
+            var settingsFile = await properties.GetSettingsFileAsync();
+            if (!string.IsNullOrEmpty(settingsFile)) {
+                return settingsFile.MakeAbsolutePathFromRRelative(projectFolder);
+            }
+            return null;
+        }
+
+        private void ReleaseSettings(string projectPath, bool save) {
+            _semaphore.Wait();
+            try {
+                var access = _settings[projectPath];
+                string settingsFile = null;
+                if (string.IsNullOrEmpty(access.Settings.SourceFile)) {
+                    settingsFile = Path.Combine(projectPath, "Settings.R");
+                }
+                if (save) {
+                    access.Settings.Save(settingsFile);
+                }
+                _settings.Remove(projectPath);
+            } finally {
+                _semaphore.Release();
+            }
+        }
+
+        class SettingsAccess : IProjectConfigurationSettingsAccess {
+            private readonly ProjectConfigurationSettingsProvider _provider;
+            private readonly string _projectPath;
+            private bool _changed;
+
+            public SettingsAccess(ProjectConfigurationSettingsProvider provider, string projectPath, ConfigurationSettingCollection settings) {
+                _provider = provider;
+                _projectPath = projectPath;
+                Settings = settings;
+                settings.CollectionChanged += OnCollectionChanged;
+            }
+
+            private void OnCollectionChanged(object sender, NotifyCollectionChangedEventArgs e) {
+                _changed = true;
+            }
+
+            #region IProjectConfigurationSettingsAccess
+            public ConfigurationSettingCollection Settings { get; }
+            public void Dispose() {
+                if (UseCount > 0) {
```

Comment:
It would have to be ```CountdownDisposable<T>``` since I need disposable object with more methods than just plain IDisposable. Or unseal it and derive SettingsAccess from it